android-project/app/src/main/java/org/diasurgical/devilutionx/DevilutionXSDLActivity.java

```diff
@@ -1,136 +1,155 @@
+			showInstallationGuide();
+		}
+
+		super.onCreate(savedInstanceState);
+	}
+
+	/**
+	 * Check if this is the first time the app is run and the full game data is not already present
+	 */
+	private boolean shouldShowInstallationGuide() {
+		mPrefs = PreferenceManager.getDefaultSharedPreferences(this);
+		return mPrefs.getBoolean(shouldShowInstallationGuidePref, !hasFullGameData());
+	}
+
+	private boolean hasFullGameData() {
+		File fileLower = new File(externalDir + "/diabdat.mpq");
+		File fileUpper = new File(externalDir + "/DIABDAT.MPQ");
+
+		return fileUpper.exists() || fileLower.exists();
+	}
+
+	private void installationGuideShown() {
+		SharedPreferences.Editor editor = mPrefs.edit();
+		editor.putBoolean(shouldShowInstallationGuidePref, false);
+		editor.apply();
+	}
+
+	private void showInstallationGuide() {
+		String message = "To play the full version you must place DIABDAT.MPQ from the original game in Android"
+				+ externalDir.split("/Android")[1]
+				+ ".\n\nIf you don't have the original game then you can buy Diablo from GOG.com.";
+		new AlertDialog.Builder(this)
+				.setTitle(SharewareTitle)
+				.setPositiveButton("OK", null)
+				.setMessage(message)
+				.show();
+		installationGuideShown();
+	}
+
+	private boolean copyFile(File src, File dst) {
+		try {
+			try (InputStream in = new FileInputStream(src)) {
+				try (OutputStream out = new FileOutputStream(dst)) {
+					// Transfer bytes from in to out
+					byte[] buf = new byte[1024];
+					int len;
+					while ((len = in.read(buf)) > 0) {
+						out.write(buf, 0, len);
+					}
+
+					return true;
+				}
+			}
+		} catch (IOException exception) {
+			Log.v("copyFile", exception.getMessage());
+		}
+
+		return false;
+	}
+
+	private void migrateFile(File file) {
+		if (!file.exists() || !file.canRead()) {
+			return;
+		}
+		File newPath = new File(externalDir + "/" + file.getName());
+		if (newPath.exists()) {
+			if (file.canWrite()) {
+				file.delete();
+			}
+			return;
+		}
+		if (!new File(newPath.getParent()).canWrite()) {
+			return;
+		}
+		if (!file.renameTo(newPath)) {
+			if (copyFile(file, newPath) && file.canWrite()) {
+				file.delete();
+			}
+		}
+	}
+
+	/**
+	 * This can be removed Nov 2021 and Google will no longer permit access to the old folder from that point on
+	 */
+	private void migrateAppData() {
+		if (Build.VERSION.SDK_INT > Build.VERSION_CODES.LOLLIPOP_MR1) {
+			if (PackageManager.PERMISSION_GRANTED != checkSelfPermission(Manifest.permission.READ_EXTERNAL_STORAGE)) {
+				return;
+			}
+		}
+
+		migrateFile(new File(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS) + "/diabdat.mpq"));
+		migrateFile(new File(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS) + "/DIABDAT.MPQ"));
+
+		migrateFile(new File("/sdcard/diabdat.mpq"));
+		migrateFile(new File("/sdcard/devilutionx/diabdat.mpq"));
+		migrateFile(new File("/sdcard/devilutionx/spawn.mpq"));

```

Comment:
Is there a reason why we're copying `DIABDAT.MPQ` from `Downloads`, but not from `/sdcard/devilutionx`?

Reply:
The previous build where hard coded for `/sdcard/devilutionx/diabdat.mpq`, if the user placed the file as `DIABDAT.MPQ` it would either not have worked or the SD card was formatted as FAT32 (case insensitive).

Reply:
I see now. Mine was the latter. It's not copying, though, because `file.canRead()` is returning `false`. I checked permissions, and my only options under "Files and media" are "Allow access to media only" or "Deny".

Reply:
I have adjusted the SDK version in the gradle file, please try again :)

Reply:
That did it.